File: src/main/java/ch/qligier/jetbrains/plugin/fhir/fsh/psi/entity/FshPsiAlias.java
// Copyright 2024 Quentin Ligier. Use of this source code is governed by the MIT license.

package ch.qligier.jetbrains.plugin.fhir.fsh.psi.entity;

import ch.qligier.jetbrains.plugin.fhir.fsh.FshLanguage;
import ch.qligier.jetbrains.plugin.fhir.fsh.grammar.FshLexer;
import ch.qligier.jetbrains.plugin.fhir.fsh.specification.ItemType;
import com.intellij.lang.ASTNode;
import org.antlr.intellij.adaptor.lexer.PSIElementTypeFactory;
import org.jspecify.annotations.Nullable;


/**
 * PSI element representing an alias in an FSH file.
 *
 * @author Quentin Ligier
 **/
public class FshPsiAlias extends FshPsiItem {

    public FshPsiAlias(final ASTNode node) {
        super(node);
    }

    @Override
    public ItemType getItemType() {
        return ItemType.ALIAS;
    }

    @Override
    public @Nullable String getItemName() {
        return this.getNode().getText(); //TODO get from name
    }

    public @Nullable String getAliasValue() {
        final var type = PSIElementTypeFactory.getTokenIElementTypes(FshLanguage.INSTANCE).get(FshLexer.SEQUENCE);
        final var child = this.getNode().findChildByType(type);
        return (child != null) ? child.getText() : null;
    }

    @Override
    public @Nullable String toString() {
        return "FshPsiAlias(%s, %s)".formatted(this.getItemName(), this.getAliasValue());
    }
}

File: src/main/kotlin/ch/qligier/jetbrains/plugin/fss/fsh/psi/entity/FshPsiAlias.kt
// Copyright 2025 Quentin Ligier. Use of this source code is governed by the MIT license.
package ch.qligier.jetbrains.plugin.fss.fsh.psi.entity

import ch.qligier.jetbrains.plugin.fss.fsh.FshLanguage
import ch.qligier.jetbrains.plugin.fss.fsh.antlr.FSHLexer
import ch.qligier.jetbrains.plugin.fss.fsh.specs.ItemType
import com.intellij.lang.ASTNode
import org.antlr.intellij.adaptor.lexer.PSIElementTypeFactory

/**
 * PSI element representing an alias in an FSH file.
 */
class FshPsiAlias(node: ASTNode) : FshPsiItem(node) {
    override fun getItemType(): ItemType {
        return ItemType.ALIAS
    }

    override fun getItemName(): String? {
        return this.node.text //TODO get from name
    }

    val aliasValue: String?
        get() {
            val type =
                PSIElementTypeFactory.getTokenIElementTypes(FshLanguage)[FSHLexer.SEQUENCE]
            val child = this.node.findChildByType(type)
            return child?.text
        }

    override fun toString(): String {
        return "FshPsiAlias(${this.getItemName()}, ${this.aliasValue})"
    }
}
